File: frontend/src/lib/components/definitions/xovi.ts
import { ComponentDefinition } from '../types'
import { CommandBuilders, ArchUtils } from '../utils'

export const xoviComponent: ComponentDefinition = {
  id: 'xovi',
  name: 'Xovi',
  description: 'Base framework for reMarkable modifications',
  version: 'v16-14112025',
  author: 'asivery',

  dependencies: [],

  checkInstalled: () => CommandBuilders.testExists('/home/root/xovi', 'directory'),

  install: (ctx) => {
    const archTesting = ArchUtils.getArchiveArch(ctx.arch)
    const baseUrl = 'https://github.com/asivery/rm-xovi-extensions/releases/download/v16-14112025'

    return [
      CommandBuilders.changeDirectory('/home/root'),
      CommandBuilders.conditional(
        'test -f curl',
        'echo "curl already exists"',
        'wget -q https://github.com/moparisern/remarkable-curl/releases/download/v0.1/curl-reMarkable2 -O curl && chmod +x curl'
      ),
      {
        script: `./curl -Lo extensions.zip ${baseUrl}/extensions-${archTesting}.zip`,
        description: 'Download xovi extensions',
      },
      {
        script: 'unzip -o -d extensions extensions.zip && rm extensions.zip',
        description: 'Extract extensions',
      },
      {
        script: 'bash extensions/install-xovi-for-rm',
        description: 'Run xovi installer',
      },
    ]
  },

  preUninstall: {
    type: 'custom',
    execute: async () => ({
      command: {
        script: '/home/root/xovi/stock',
        description: 'Restore device to stock state',
      },
    }),
  },

  uninstall: () => [
    CommandBuilders.removeDirectory('/home/root/xovi'),
    CommandBuilders.removeDirectory('/home/root/extensions'),
  ],

  enable: () => ({
    script: '/home/root/xovi/start',
    description: 'Start xovi',
  }),

  disable: () => ({
    script: '/home/root/xovi/stop',
    description: 'Stop xovi',
  }),

  maintenanceCommands: [
    {
      id: 'start',
      label: 'Start Xovi',
      description: 'Start the xovi framework',
      command: () => ({ script: '/home/root/xovi/start', description: 'Start xovi' }),
    },
    {
      id: 'debug',
      label: 'Debug Mode',
      description: 'Run xovi in debug mode with live output',
      command: () => ({ script: '/home/root/xovi/debug', description: 'Run xovi debug' }),
      requiresTerminal: true,
      allowStop: true,
    },
    {
      id: 'stock',
      label: 'Restore Stock',
      description: 'Restore device to stock remarkable software',
      command: () => ({ script: '/home/root/xovi/stock', description: 'Restore stock' }),
      requiresTerminal: true,
    },
  ],

  category: 'core',
  tags: ['framework', 'required'],
  requiresReboot: false,
}

File: internal/component/definitions/xovi.go
package definitions

import (
	"fmt"

	"reManager/internal/commands"
	"reManager/internal/component"
)

var XoviComponent = &component.ComponentDefinition{
	ID:          "xovi",
	Name:        "Xovi",
	Description: "Base framework for reMarkable modifications",
	Version:     "v16-14112025",
	Author:      "asivery",

	Dependencies: []string{},

	CheckInstalled: func(ctx component.CommandContext) component.CommandResult {
		return commands.TestExists("/home/root/xovi", commands.PathDirectory)
	},

	Install: func(ctx component.CommandContext) []component.CommandResult {
		archTesting := commands.GetArchiveArch(ctx.Arch)
		baseUrl := "https://github.com/asivery/rm-xovi-extensions/releases/download/v16-14112025"

		return []component.CommandResult{
			commands.ChangeDirectory("/home/root"),
			commands.Conditional(
				"test -f curl",
				"echo \"curl already exists\"",
				"wget -q https://github.com/moparisern/remarkable-curl/releases/download/v0.1/curl-reMarkable2 -O curl && chmod +x curl",
			),
			{
				Script:      fmt.Sprintf("./curl -Lo extensions.zip %s/extensions-%s.zip", baseUrl, archTesting),
				Description: "Download xovi extensions",
			},
			{
				Script:      "unzip -o -d extensions extensions.zip && rm extensions.zip",
				Description: "Extract extensions",
			},
			{
				Script:      "bash extensions/install-xovi-for-rm",
				Description: "Run xovi installer",
			},
		}
	},

	PreUninstall: &component.LifecycleHook{
		Type: component.HookTypeCustom,
		Execute: func(ctx component.CommandContext) (*component.HookExecutionResult, error) {
			return &component.HookExecutionResult{
				Command: &component.CommandResult{
					Script:      "/home/root/xovi/stock",
					Description: "Restore device to stock state",
				},
			}, nil
		},
	},

	Uninstall: func(ctx component.CommandContext) []component.CommandResult {
		return []component.CommandResult{
			commands.RemoveDirectory("/home/root/xovi"),
			commands.RemoveDirectory("/home/root/extensions"),
		}
	},

	Enable: func(ctx component.CommandContext) *component.CommandResult {
		return &component.CommandResult{
			Script:      "/home/root/xovi/start",
			Description: "Start xovi",
		}
	},

	Disable: func(ctx component.CommandContext) *component.CommandResult {
		return &component.CommandResult{
			Script:      "/home/root/xovi/stop",
			Description: "Stop xovi",
		}
	},

	MaintenanceCommands: []component.MaintenanceCommand{
		{
			ID:          "start",
			Label:       "Xovi Start",
			Description: "Start the xovi framework",
			Command: func(ctx component.CommandContext) []component.CommandResult {
				return []component.CommandResult{{Script: "/home/root/xovi/start", Description: "Start xovi"}}
			},
		},
		{
			ID:               "debug",
			Label:            "Xovi Debug",
			Description:      "Run xovi in debug mode with live output",
			RequiresTerminal: true,
			AllowStop:        true,
			Command: func(ctx component.CommandContext) []component.CommandResult {
				return []component.CommandResult{{Script: "/home/root/xovi/debug", Description: "Run xovi debug", RequiresPTY: true}}
			},
		},
		{
			ID:               "stock",
			Label:            "Xovi Stock",
			Description:      "Restore device to stock remarkable software",
			RequiresTerminal: true,
			Command: func(ctx component.CommandContext) []component.CommandResult {
				return []component.CommandResult{{Script: "/home/root/xovi/stock", Description: "Restore stock"}}
			},
		},
	},

	Category:       "core",
	Tags:           []string{"framework", "required"},
	RequiresReboot: false,
}
